src/Workspaces/Core/Portable/RegularExpressions/RegexPatternDetector.cs

```diff
@@ -0,0 +1,353 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.RegularExpressions
+{
+    /// <summary>
+    /// Helper class to detect regex pattern tokens in a document efficiently.
+    /// </summary>
+    internal class RegexPatternDetector
+    {
+        private const string _patternName = "pattern";
+
+        private static readonly ConditionalWeakTable<SemanticModel, RegexPatternDetector> _modelToDetector =
+            new ConditionalWeakTable<SemanticModel, RegexPatternDetector>();
+
+        private readonly SemanticModel _semanticModel;
+        private readonly ISyntaxFactsService _syntaxFacts;
+        private readonly ISemanticFactsService _semanticFacts;
+        private readonly INamedTypeSymbol _regexType;
+        private readonly HashSet<string> _methodNamesOfInterest;
+
+        /// <summary>
+        /// Helps match patterns of the form: language=regex,option1,option2,option3
```

Comment:
@kuhlenh If we add this feature, it would be good to doc this format.

Specifically, this allows users to write arbitrary strings, but annotate them with a comment so that the IDE will treat the string as a regex.  i.e. you can write:

```c#
private readonly string regexText = /* language=regex,explicitcature */ "...";
```

And all the IDE will perform all the regex features in it.